packages/@aws-cdk/aws-elasticsearch/lib/domain.ts

```diff
@@ -0,0 +1,1271 @@
+   *
+   * @default - false
+   */
+  readonly slowIndexLogEnabled?: boolean;
+
+  /**
+   * Log slow indices to this log group.
+   *
+   * @default - a new log group is created if slow index logging is enabled
+   */
+  readonly slowIndexLogGroup?: logs.ILogGroup;
+
+  /**
+   * Specify if Elasticsearch application logging should be set up.
+   * Requires Elasticsearch version 5.1 or later.
+   *
+   * @default - false
+   */
+  readonly appLogEnabled?: boolean;
+
+  /**
+   * Log Elasticsearch application logs to this log group.
+   *
+   * @default - a new log group is created if app logging is enabled
+   */
+  readonly appLogGroup?: logs.ILogGroup;
+}
+
+/**
+ * Whether the domain should encrypt data at rest, and if so, the AWS Key
+ * Management Service (KMS) key to use. Can only be used to create a new domain,
+ * not update an existing one. Requires Elasticsearch version 5.1 or later.
+ */
+export interface EncryptionAtRestOptions {
+  /**
+   * Specify true to enable encryption at rest.
+   *
+   * @default - encryption at rest is disabled.
+   */
+  readonly enabled?: boolean;
+
+  /**
+   * Supply if using KMS key for encryption at rest.
+   *
+   * @default - uses default aws/es KMS key.
+   */
+  readonly kmsKey?: kms.IKey;
+}
+
+/**
+ * Configures Amazon ES to use Amazon Cognito authentication for Kibana.
+ */
+export interface CognitoOptions {
+  /**
+   * The Amazon Cognito identity pool ID that you want Amazon ES to use for Kibana authentication.
+   */
+  readonly identityPoolId: string;
+
+  /**
+   * The AmazonESCognitoAccess role that allows Amazon ES to configure your user pool and identity pool.
+   */
+  readonly role: iam.IRole;
+
+  /**
+   * The Amazon Cognito user pool ID that you want Amazon ES to use for Kibana authentication.
+   */
+  readonly userPoolId: string;
+}
+
+/**
+ * The virtual private cloud (VPC) configuration for the Amazon ES domain. For
+ * more information, see [VPC Support for Amazon Elasticsearch Service
+ * Domains](https://docs.aws.amazon.com/elasticsearch-service/latest/developerguide/es-vpc.html)
+ * in the Amazon Elasticsearch Service Developer Guide.
+ */
+export interface VpcOptions {
+  /**
+   * The list of security groups that are associated with the VPC endpoints
+   * for the domain. If you don't provide a security group ID, Amazon ES uses
+   * the default security group for the VPC. To learn more, see [Security Groups for your VPC]
+   * (https://docs.aws.amazon.com/vpc/latest/userguide/VPC_SecurityGroups.html) in the Amazon VPC
+   * User Guide.
+   */
+  readonly securityGroups: ec2.ISecurityGroup[];
+
+  /**
+   * Provide one subnet for each Availability Zone that your domain uses. For
+   * example, you must specify three subnet IDs for a three Availability Zone
+   * domain. To learn more, see [VPCs and Subnets]
+   * (https://docs.aws.amazon.com/vpc/latest/userguide/VPC_Subnets.html) in the
+   * Amazon VPC User Guide.
+   */
+  readonly subnets: ec2.ISubnet[];
```

Comment:
The VPC has a lot more configuration than what ES accepts https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-ec2.IVpc.html. I don't want to give the user the impression that those extra settings are somehow applied when they aren't. Also do I use the private or public subnets from the VPC?

Reply:
@iliapolo @stephanh I'll wait to address this feedback until the conversation resolves.

Reply:
> The VPC has a lot more configuration than what ES accepts

Thats ok, we are not requiring the user to provide all of these, nor we are passing them to the cluster configuration in any way. The idea of accepting a VPC is to do automatic subnet selection, and provision the cluster in its own VPC by default. 

As for the default subnets, lets just select all of them by default. Note that for multi zone clusters, the docs say:

> If you enable multiple Availability Zones for your domain, each subnet must be in a different Availability Zone in the same region. 

So if selecting all subnets fail in this scenario, we can be smarter by using `[{onePerAz: true}]` as the default subnet selection

Reply:
Just to confirm, based on the hoisting comment below, we're going to end up with the following additional options on `DomainProps`:
```
readonly vpc?: ec2.IVpc;
readonly vpcSubnets?: ec2.SubnetSelection[];
readonly securityGroups?: ec2.ISecurityGroup[];
```

Also, if `vpc` should default to creating a new VPC for the Domain, how do we want to allow for a non-VPC (internet endpoint) Domain?

Reply:
Placing the domain in a VPC doesn't necessarily mean that it's not accessible from the internet. If a VPC has a public subnet, and that subnet is attached to the domain, than its accessible as far as networking is concerned. 

Note that the domain has to be attached a subnet, and a subnet has to reside inside a VPC (might be the default one created for that region - but still a VPC), is there is effectively no such thing as a VPCless domain.

Reply:
If that's the case, then [this documentation](https://docs.aws.amazon.com/elasticsearch-service/latest/developerguide/es-vpc.html) is confusing to me:

> Limitations
> Currently, operating an Amazon ES domain within a VPC has the following limitations:
> 
> You can either launch your domain within a VPC or use a public endpoint, but you can't do both. You must choose one or the other when you create your domain.
> 
> If you launch a new domain within a VPC, you can't later switch it to use a public endpoint. The reverse is also true: If you create a domain with a public endpoint, you can't later place it within a VPC. Instead, you must create a new domain and migrate your data.

Reply:
@iliapolo when going through the console and creating ES domains, there seems to be a pretty clear distinction between VPC and internet/public access. Given that `vpcOptions` is optional from a CF standpoint, are you sure we should be creating a VPC if the user doesn't provide one?

<img width="1114" alt="Screen Shot 2020-08-18 at 1 50 25 PM" src="https://user-images.githubusercontent.com/2363879/90553299-c6b04080-e159-11ea-9f36-bc21b2dad917.png">

Reply:
@adamelmore - Yeah, I was just playing with it as well. What threw me off initially was the fact that you made `subnets ` required, but it's only required if `vpcOptions` is specified...sorry about the mixup.

Pretty sure you can still make the domain public by specifying a VPC and a public subnet though. Anyway, I do eventually want us to create a VPC on the user's behalf, but since I can't come up with a nice API for it right now, i'm ok with rolling this out as is.